pss/psstypes.py

```diff
@@ -91,26 +92,50 @@ def parser(
         min=None,
         max=None,
         parent=None,
-        choices=None
+        choices=None,
+        transform=None
 ):
-    def inner(func):
+    '''We want to be able to call the parser as both a
+    decorator and call it without including a function.
+
+    We initially set the `type_name` parser to be an identity
+    function (or the passed in `transform` function). If this
+    function is being used as a decorator, we overwrite the
+    parser with the decorated function.
+
+    Additionally, we use a closure so we can validate (check
+    the regex, min, max, choices, etc.) the output of
+    whatever function is used for the parser.
+
+    NOTE: the `transform` parameter will do nothing (gets
+    overwritten) when this function is called as a decorator
+    '''
+    def validate_value(value_function):
         def new_func(value, **kwargs):
             if validation_regexp and isinstance(value, str):
                 if not re.match(validation_regexp, value):
                     raise ValueError(f"Value '{value}' does not match the required pattern {validation_regexp})")
             if parent:
                 value = _TYPES_DICT[parent]['parser'](value)
-            parsed = func(value, **kwargs)
+            parsed = value_function(value, **kwargs)
             if min is not None and parsed < min:
                 raise ValueError(f"Value '{value}' ('{parsed}') is less than {min}")
             if max is not None and parsed > max:
                 raise ValueError(f"Value '{value}' ('{parsed}') is more than {max}")
             if choices is not None and value not in choices:
                 raise ValueError(f"Value '{value}' ('{parsed}') is not a valid option for {type_name}. Available choices: {choices}")
             return parsed
-
-        _TYPES_DICT[type_name]['parser'] = new_func
         return new_func
+
+    def inner(func):
+        _TYPES_DICT[type_name]['parser'] = validate_value(func)
+        @functools.wraps(func)
+        def wrapper(*args, **kwargs):
+            return func(*args, **kwargs)
+        return wrapper
+
+    transformation_func = transform if callable(transform) else lambda val: val
```

Comment:
This should be `transform is not None`. If we pass a number or string to this, it should raise an exception about a bad argument, and not default to an identity function. 

I would actually go one step further, and default `transform` to a sentinel. We should expect either:

* `transform` is a function
* This is used as a decorator
* `transform` is set to `None` (in which case, it defaults to the identity). This could even be a local sentinel-style `pmss.NONE`.